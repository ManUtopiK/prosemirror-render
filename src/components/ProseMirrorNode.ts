import {Component, computed, defineComponent, h, PropType, resolveComponent, toRefs} from "vue";
import { snakeCase, kebabCase } from "change-case";
import type {Base, Node} from "../prosemirror-json.ts";

import { defaultConfig as config} from "../config.ts";

/*
interface ProseMirrorNodeProperties {
    // curent prosemirror node
    node: Node;
    // mark index to render
    mark?: number;
}
*/

function substituteAttributes(name: string, attrs?: Record<string, any>): string {
    const regex = /\[([a-zA-Z_]\w*)]/g;

    return name.replace(regex, (_match: string, variable: string) => {
        return attrs ? attrs[variable] ?? variable : variable;
    });
}

const resolveProseComponent = (el: Base, typeMap: Record<string, string | Component>) => {
    const type = snakeCase(el.type);
    const name: string | Component = typeMap[type] ?? "prose-mirror-" + kebabCase(el.type);
    const component: string | Component = typeof name === "string" ? resolveComponent(name) : name;
    const parsed: string | Component = typeof component === "string" ? substituteAttributes(component, el.attrs) : component;

    return parsed;
}

const ProseMirrorNode = defineComponent({
    name: "ProseMirrorNode",
    props: {
        node: { type: Object as PropType<Node>, required: true },
        mark: Number,
    },
    setup(properties) {
        const me= resolveComponent("ProseMirrorNode", true);

        const { typeMap } = config;

        const { node, mark } = toRefs(properties);

        // point to the mark
        const markIndex = computed(() => mark.value ?? 0);
        const markItem = computed(() => node.value.marks?.at(markIndex.value));

        return () => {
            // render the current mark
            if (markItem.value) {
                const markComponent = resolveProseComponent(markItem.value, typeMap);
                return h(
                    markComponent,
                    markItem.value.attrs,
                    // recurse the next mark for child
                    h(me, { node: node.value, mark: markIndex.value + 1 }),
                );
            }
            // render text as is
            else if (node.value.type === "text") {
                return node.value.text;
            }
            // render the current node when marks are done
            else {
                const proseComponent = resolveProseComponent(node.value, typeMap);
                return h(
                    proseComponent,
                    { ...node.value.attrs, node: node.value },
                    // node content build the children
                    node.value.content?.map((child) => h(me, { node: child })),
                );
            }
        };
    },
});

export default ProseMirrorNode;
