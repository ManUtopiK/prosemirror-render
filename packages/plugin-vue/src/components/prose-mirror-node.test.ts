import { mount } from "@vue/test-utils";
import ProseMirrorNode, { resolveProseComponent, substituteAttributes } from "./prose-mirror-node.ts";
import { describe, it, expect } from "vitest";

describe("component ProseMirrorNode", () => {
  it("renders simple node", async () => {
    expect(ProseMirrorNode).toBeTruthy();

    const wrapper = mount(ProseMirrorNode, {
      props: {
        node: {
          type: "doc",
          content: [
            {
              type: "paragraph",
              content: [
                {
                  type: "text",
                  text: "content",
                },
              ],
            },
          ],
        },
      },
    });

    expect(wrapper.text()).toContain("content");
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("renders simple mark", async () => {
    expect(ProseMirrorNode).toBeTruthy();

    const wrapper = mount(ProseMirrorNode, {
      props: {
        node: {
          type: "doc",
          content: [
            {
              type: "paragraph",
              content: [
                {
                  type: "text",
                  marks: [
                    {
                      type: "bold",
                    },
                  ],
                  text: "content",
                },
              ],
            },
          ],
        },
      },
    });

    expect(wrapper.text()).toContain("content");
    expect(wrapper.html()).toContain("<b>content</b>");
    expect(wrapper.html()).toMatchSnapshot();
  });
});

describe("substituteAttributes", () => {
  it("replaces placeholders", () => {
    expect(substituteAttributes("paragraph[foo]")).toBe("paragraphfoo");
    expect(substituteAttributes("paragraph[foo]", { foo: "bar" })).toBe("paragraphbar");

    expect(substituteAttributes("h[level]", { level: 1 })).toBe("h1");
  });

  it("noop if no placeholder", () => {
    expect(substituteAttributes("h", { level: 1 })).toBe("h");
  });
});

describe("resolveProseComponent", () => {
  it("returns the element name", () => {
    expect(resolveProseComponent({ type: "paragraph" }, { paragraph: "p" })).toBe("p");
  });

  it("returns the node type if no correspondance", () => {
    expect(resolveProseComponent({ type: "paragraph" }, {})).toBe("paragraph");
  });

  it("finds the type in camel case", () => {
    expect(resolveProseComponent({ type: "camelCase" }, { camel_case: "p" })).toBe("p");
  });

  it("finds the type in pascal case", () => {
    expect(resolveProseComponent({ type: "kebab-case" }, { kebab_case: "p" })).toBe("p");
  });

  it("returns a component", () => {
    expect(resolveProseComponent({ type: "p" }, { p: ProseMirrorNode })).toBe(ProseMirrorNode);
  });

  it("tries to resolve to a component", () => {
    expect(resolveProseComponent({ type: "foo-bar" }, {})).toBe("foo-bar");
  });
});
