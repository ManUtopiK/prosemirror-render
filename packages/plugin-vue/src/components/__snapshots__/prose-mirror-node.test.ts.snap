// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`component ProseMirrorNode > renders simple mark 1`] = `
"<div node=\\"[object Object]\\">
  <p node=\\"[object Object]\\"><b>content</b></p>
</div>"
`;

exports[`component ProseMirrorNode > renders simple node 1`] = `
"<div node=\\"[object Object]\\">
  <p node=\\"[object Object]\\">content</p>
</div>"
`;
